### PR TITLE
VultureBear: Make use of vulture API

### PR DESCRIPTION
**Pull request by RJ722**
### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bear-requirements.txt [hunk 1]
```diff
@@ -34,6 +34,6 @@ rstcheck~=3.1
 safety~=0.5.1
 scspell3k~=2.0
 vim-vint~=0.3.12
-vulture~=0.14.0
+vulture==0.24.0
```
Comment (jendrikseipp) [inline bear-requirements.txt]: Can we use 0.25? It should make integration even easier.

File: tests/python/VultureBearTest.py [hunk 2]
```diff
@@ -42,85 +42,116 @@ def prep_file():
...
+        print("Unreachable")
+"""
+        self.verify_results(bad_file, {
+            "unsatisfiable 'if' condition": (2, 3, 100)
+        })
+
+    def test_unreachable_else(self):
+        bad_file = """\
+if True:
+    print("Reachanle")
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: typo

File: tests/python/VultureBearTest.py [hunk 3]
```diff
@@ -42,85 +42,116 @@ def prep_file():
...
+                                        item.affected_code[0].end.line,
+                                        item.confidence))
+                        for item in self.get_results(test_case))
+        self.assertEqual(detected, expected)
+
     def test_used_variable(self):
-        good_file = """
-        x = 2
-        print(x)
+        good_file = """\
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: instead of `good_file` and `bad_file`, can we just use `code`?

**Comment by jendrikseipp** [comment]
Congratulations on getting PR number 2000 :-)

File: tests/python/VultureBearTest.py (hunk 3 above)
Comment (RJ722) [inline tests/python/VultureBearTest.py]: Yeah! That indeed looks  irrelevant! 🤔

File: bear-requirements.txt (hunk 1 above)
Comment (RJ722) [inline bear-requirements.txt]: The new `Item` class is so cool. Native support for `coala` ;-)

**Comment by Nosferatul** [comment]
i think this looks good right now, what do you say @jendrikseipp ?

**Comment by jendrikseipp** [comment]
I'll have a look :-)

File: bears/python/VultureBear.py [hunk 4]
```diff
@@ -3,42 +3,26 @@
...
-                                 confidence=CONFIDENCE_MAP[item.typ])
+                                 line=item.first_lineno,
+                                 end_line=item.last_lineno,
+                                 confidence=item.confidence)
 
 
 class VultureBear(GlobalBear):
     LANGUAGES = {'Python', 'Python 3'}
-    REQUIREMENTS = {PipRequirement('vulture', '0.14.0')}
+    REQUIREMENTS = {PipRequirement('vulture', '0.25.0')}
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: Is it normal that we have to list the requirement in two different places?

File: tests/python/VultureBearTest.py [hunk 5]
```diff
@@ -42,85 +42,116 @@ def prep_file():
...
-        self.assertEqual(len(result), 2)
-        self.assertEqual(result[0].confidence, 70)
-
-    def test_func_confidence(self):
-        bad_file = """
-        def hello(name):
-            print('Hello World')
+    def __init__(self, name):
+        self.name = name
+        self.sur_name = 'Something'
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: sur_name -> surname

File: tests/python/VultureBearTest.py [hunk 6]
```diff
@@ -42,85 +42,116 @@ def prep_file():
...
+        print("Unreachable")
+"""
+        self.verify_results(code, {
+            "unsatisfiable 'if' condition": (2, 3, 100)
+        })
+
+    def test_unreachable_else(self):
+        code = """\
+if True:
+    print("Reachanle")
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: typo

File: tests/python/VultureBearTest.py [hunk 7]
```diff
@@ -42,85 +42,116 @@ def prep_file():
...
+        self.verify_results(code, {
+            "unused import 'subprocess'": (2, 2, 90)
+        })
 
-    def test_class_confidence(self):
-        bad_file = """
-        class Name:
+    def test_class(self):
+        code = """\
+class Name:
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: Can you please remove this empty line?

File: bears/python/VultureBear.py [hunk 8]
```diff
@@ -3,42 +3,26 @@
...
-
-    vulture = Vulture()
+    vulture = Vulture(verbose=True)
     vulture.scavenge(filenames)
-    for item in sorted(
-            vulture.unused_funcs + vulture.unused_imports +
-            vulture.unused_props + vulture.unused_vars +
-            vulture.unused_attrs, key=file_lineno):
-        message = 'Unused {0}: {1}'.format(item.typ, item)
+    for item in vulture.get_unused_code(sort_by_size=True):
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: Is sorting helpful here? I would guess that coala has its own sorting. And if it doesn't, I think we should sort by filenames.

File: bears/python/VultureBear.py [hunk 9]
```diff
@@ -3,42 +3,26 @@
...
         yield Result.from_values(origin='VultureBear',
-                                 message=message,
+                                 message=item.message,
                                  file=item.filename,
-                                 line=item.lineno,
-                                 confidence=CONFIDENCE_MAP[item.typ])
+                                 line=item.first_lineno,
+                                 end_line=item.last_lineno,
+                                 confidence=item.confidence)
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: Is it already possible to pass a minimum confidence to coala? If not, we should create an issue for this.

File: bears/python/VultureBear.py [hunk 10]
```diff
@@ -3,42 +3,26 @@
...
     """
     :param filenames: List of filenames to check.
     :return: Generator of Result objects.
     """
-
-    def file_lineno(item):
-        return (item.filename.lower(), item.lineno)
-
-    vulture = Vulture()
+    vulture = Vulture(verbose=True)
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: I don't think we should be verbose.

File: bears/python/VultureBear.py (hunk 9 above)
Comment (RJ722) [inline bears/python/VultureBear.py]: I don't know!

CC @jayvdb

File: bears/python/VultureBear.py (hunk 4 above)
Comment (RJ722) [inline bears/python/VultureBear.py]: I remember doing it previously. FWIW I have asked in the gitter channel.

File: bears/python/VultureBear.py (hunk 8 above)
Comment (RJ722) [inline bears/python/VultureBear.py]: I think coala doesn't sorts the results at all because results are yielded at runtime directly, without being stored (generally).

Also, since files would be automatically be coming in an alphabetical order, I think we won't need additional sorting. I am removing the `sort_by_size=True` for now. Please comment if you think else wise. :-)

File: tests/python/VultureBearTest.py [hunk 11]
```diff
@@ -42,85 +42,115 @@ def prep_file():
...
-        class Bar(object):
-            @property
-            def prop(self):
-                pass
-
-        c = Bar()
+        self.verify_results(code, {
+            "unused class 'Name'": (1, 5, 60),
+            "unused attribute 'name'": (4, 4, 60),
+            "unused attribute 'surname'": (5, 5, 60)
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: Aren't these line numbers outdated?

Comment (RJ722) [inline tests/python/VultureBearTest.py]: sorry, I didn't run tests. Corrected!

**Comment by jendrikseipp** [comment]
The code looks good. This can be merged once all tests pass from my point of view.

File: bears/python/VultureBear.py (hunk 9 above)
Comment (Makman2) [inline bears/python/VultureBear.py]: we plan to support this, I think there should be somewhere an issue :+1: (can't find it right now)

Comment (Makman2) [inline bears/python/VultureBear.py]: otherwise feel free to file one ;)

File: bears/python/VultureBear.py (hunk 4 above)
Comment (Makman2) [inline bears/python/VultureBear.py]: yeah currently we need it at two places, one for CI and one for the bear itself. Once our requirement-management gets more mature we will generate that dynamically as far as possible :+1:

File: tests/python/VultureBearTest.py [hunk 12]
```diff
@@ -42,85 +42,115 @@ def prep_file():
...
+        self.verify_results(code, {
+            "unused import 'subprocess'": (2, 2, 90)
+        })
+
+    def test_class(self):
+        code = """\
+class Name:
+    def __init__(self, name):
+        self.name = name
+        self.surname = 'Something'
```
Comment (Makman2) [inline tests/python/VultureBearTest.py]: please make test files for these, the code is horrible to read having this inlined^^

File: bears/python/VultureBear.py [hunk 13]
```diff
@@ -3,42 +3,26 @@
...
-        message = 'Unused {0}: {1}'.format(item.typ, item)
+    for item in vulture.get_unused_code():
         yield Result.from_values(origin='VultureBear',
-                                 message=message,
+                                 message=item.message,
                                  file=item.filename,
-                                 line=item.lineno,
-                                 confidence=CONFIDENCE_MAP[item.typ])
+                                 line=item.first_lineno,
+                                 end_line=item.last_lineno,
```
Comment (Makman2) [inline bears/python/VultureBear.py]: looks like there are specific things you use now from the newer vulture version (like `item.first_lineno` or `item.confidence`). Could you list them in the commit message?

File: tests/python/VultureBearTest.py [hunk 14]
```diff
@@ -42,85 +42,115 @@ def prep_file():
 
             return list(self.uut.run())
 
+    def verify_results(self, test_case, expected):
+        detected = dict((item.message, (item.affected_code[0].start.line,
+                                        item.affected_code[0].end.line,
+                                        item.confidence))
+                        for item in self.get_results(test_case))
+        self.assertEqual(detected, expected)
```
Comment (Makman2) [inline tests/python/VultureBearTest.py]: why not use `LocalBearTestHelper.check_results`?

Comment (Makman2) [inline tests/python/VultureBearTest.py]: hm okay it's a `GlobalBear`, though can we do something similar?

File: tests/python/VultureBearTest.py (hunk 12 above)
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: I'm not sure separate files are better here. The current solution has the advantage that it keeps the code and the expected results close together. Separating them makes maintainance harder, I think.

File: bears/python/VultureBear.py (hunk 13 above)
Comment (RJ722) [inline bears/python/VultureBear.py]: Is it all right now?

**Comment by jendrikseipp** [comment]
@Makman2 Can this be merged?

File: tests/python/VultureBearTest.py (hunk 14 above)
Comment (RJ722) [inline tests/python/VultureBearTest.py]: @Makman2 IMHO `VultureBear` doesn't needs to be rigorously tested here. All it does is pass on values from vulture. Can we open a different issue and then change this afterwards?

Comment (Makman2) [inline tests/python/VultureBearTest.py]: kk fine with me :+1:

File: bears/python/VultureBear.py [hunk 15]
```diff
@@ -3,42 +3,26 @@
...
-        return (item.filename.lower(), item.lineno)
-
     vulture = Vulture()
     vulture.scavenge(filenames)
-    for item in sorted(
-            vulture.unused_funcs + vulture.unused_imports +
-            vulture.unused_props + vulture.unused_vars +
-            vulture.unused_attrs, key=file_lineno):
-        message = 'Unused {0}: {1}'.format(item.typ, item)
+    for item in vulture.get_unused_code():
```
Comment (Makman2) [inline bears/python/VultureBear.py]: does `vulture.get_unused_code()` sort the entries by line-number?

File: tests/python/VultureBearTest.py (hunk 12 above)
Comment (Makman2) [inline tests/python/VultureBearTest.py]: as Python uses indentation to identify scopes, I would definitely do this. For me this really interrupts reading flow here.
Also I haven't seen that this is much of a maintenance problem with my gsoc-student who has usually used test-files when he has written new bears :) At least in the final commit you can see the changes inside the test file and test-code side-by-side :)

Comment (Makman2) [inline tests/python/VultureBearTest.py]: This would btw avoid doing `"""\`, because I generally dislike using `\` to skip the newline :+1:

File: bears/python/VultureBear.py (hunk 15 above)
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: Yes, it does.

File: tests/python/VultureBearTest.py (hunk 12 above)
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: This change should probably be made in a separate PR though, right? The original code also uses inline code and the goal of this PR is just upgrading the Vulture version.

Comment (Makman2) [inline tests/python/VultureBearTest.py]: > This change should probably be made in a separate PR though, right?

Hm alright^^ However you could just make a commit on top if you like inside this PR. Using test files is not a time-consuming change :) See for example https://github.com/coala/coala-bears/commit/b49553a8d62cae9989d0778728adb09fecb7fffb#diff-c53fab5561bd6a2c764cf31f7aae42b2R12

Anyway what I've just seen: Some old test cases are renamed and the code is reindented. If possible keep the old tests as intact as possible. This eases review. Everything else would belong to a test-refactor which you would like to avoid in this PR and should be also put inside a separate commit/PR :+1: If the new (not-existing) indentation is now required in the code samples, we should already introduce test-files^^

File: tests/python/VultureBearTest.py [hunk 16]
```diff
@@ -42,85 +52,64 @@ def prep_file():
...
-        print(a)
-        """
-        self.assertEqual(len(self.get_results(bad_file)), 1)
-
-    def test_unused_parameter(self):
-        bad_file = """
-        def test(a):
-            return 1
-
-        test(1)
```
Comment (Makman2) [inline tests/python/VultureBearTest.py]: have you migrated tests for this case?

File: tests/python/VultureBearTest.py [hunk 17]
```diff
@@ -42,85 +52,64 @@ def prep_file():
...
-        """
-        self.assertEqual(len(self.get_results(bad_file)), 1)
-
-    def test_unused_function(self):
```
Comment (Makman2) [inline tests/python/VultureBearTest.py]: also for this one I can't find the corresponding test file :+1:

**Comment by Makman2** [comment]
I think we should squash both commits ;)

File: tests/python/VultureBearTest.py (hunk 16 above)
Comment (RJ722) [reply tests/python/VultureBearTest.py]: Yes, I just tweaked it a bit! ;-)
Here: https://github.com/coala/coala-bears/pull/2000/files#diff-2c03a6d6592124d9383a8a71826a02b6R1

File: tests/python/VultureBearTest.py (hunk 17 above)
Comment (RJ722) [reply tests/python/VultureBearTest.py]: Same here, I changed it a little bit :p 
Here: https://github.com/coala/coala-bears/pull/2000/files#diff-3e91e594a33cd00f926e16490633fadcR1

**Comment by RJ722** [comment]
Done! :smile:

**Comment by Makman2** [comment]
1. About the shortlog: I would insert the word "new", this definitely clears up some things if you don't read the commit body :) --> ` VultureBear: Make use of new vulture API`
2. You don't need an `__init__.py` file inside a test file directory, as you don't load them as a module :)

File: tests/python/vulture_test_files/unused_import.py [hunk 18]
```diff
@@ -0,0 +1,2 @@
+from os import *  # Ignore PyUnusedCodeBear
```
Comment (Makman2) [inline tests/python/vulture_test_files/unused_import.py]: Exclude those test files in our coafile :+1: Not that we use VultureBear somewhen ;)

**Comment by RJ722** [comment]
We need to silence `PythonPackageInitBear`(https://github.com/coala/coala/issues/4673)

**Comment by Makman2** [comment]
Reintroduce `__init__.py` for now, we will open another issue blocked by https://github.com/coala/coala/issues/4673 removing it again :+1:

**Comment by RJ722** [comment]
Done @Makman2 Thanks! 👍

NOTE to myself: Once this gets merged, create an issue for removing `__init__.py`.

**Comment by jayvdb** [comment]
Interesting.  Have you checked if there is a new version of alex which has new rules preventing 'his' ?

**Comment by RJ722** [comment]
@jayvdb I don't think so there were any upstream changes, https://github.com/azu/textlint-rule-alex/ was last updated - Nov, 16.

**Comment by RJ722** [comment]
Did you try rebuilding without cache?

File: tests/python/VultureBearTest.py [hunk 19]
```diff
@@ -42,85 +52,64 @@ def prep_file():
...
-        self.assertEqual(result[0].confidence, 70)
-
-    def test_var_confidence(self):
-        bad_file = """
-        name = 'Foo Bar'
-        """
-        result = self.get_results(bad_file)
-        self.assertEqual(len(result), 1)
-        self.assertEqual(result[0].confidence, 70)
+        code = load_testfile('unused_variable.py')
```
Comment (jendrikseipp) [inline tests/python/VultureBearTest.py]: I'd pass the filename to `self.verify_results()`.

Comment (RJ722) [reply tests/python/VultureBearTest.py]: Done! :+1:

**Comment by jendrikseipp** [comment]
It seems the tests are still failing. Is there something we can do about that?

**Comment by Makman2** [comment]
phew CI is going crazy... try a rebase, we recently merged some fix for CI.

**Comment by Makman2** [comment]
but yeah commits looks also fine to me :)

**Comment by jendrikseipp** [comment]
@RJ722 Does rebasing help with the tests?

**Comment by RJ722** [comment]
> @RJ722 Does rebasing help with the tests?

Done rebasing. Praying starts. 🙏

**Comment by jendrikseipp** [comment]
@RJ722 Tests passed :-) 

@Makman2 can we merge this?

**Comment by Makman2** [comment]
reack 58a3999

**Comment by Makman2** [comment]
@rultor merge

**Comment by Makman2** [comment]
yeah :boom: tough PR, but we got it :D

**Comment by RJ722** [comment]
> yeah 💥 tough PR, but we got it :D

@Makman2 Roger that!

BTW I've opened #2025

File: bears/python/VultureBear.py (hunk 4 above)
Comment (sils) [reply bears/python/VultureBear.py]: @jendrikseipp FYI the requirements.txt above is generated and the CI always checks that it's up to date so this REQUIREMENTS attribute is the correct and important place, the requirements.txt a technical necessity currently manually managed but at least enforced by the CI (last time I checked)

Comment (jendrikseipp) [reply bears/python/VultureBear.py]: @sils Thanks for clarifying!
